File: solutions/2.Add_Two_Numbers/add-two-numbers.py
# Source : https://leetcode.com/problems/add-two-numbers/
# Author : Zeming (Vantic) Zhou
# Date   : 2016-11-25

# Definition for singly-linked list.
#class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        result = ListNode(0)
        current = result
        carry = 0
        while l1 or l2:
            sumVal = carry
            if l1:
                sumVal += l1.val
                l1 = l1.next
            if l2:
                sumVal += l2.val
                l2 = l2.next
            carry = sumVal / 10
            current.val = sumVal % 10
            if l1 or l2 or carry != 0:
                current.next = ListNode(carry)
                current = current.next

        return result

File: solutions/2.Add_Two_Numbers/add-two-numbers.js
// Source : https://leetcode.com/problems/add-two-numbers/
// Author : Zeming (Vantic) Zhou
// Date   : 2016-11-25


// function ListNode(val) {
// 	this.val = val;
// 	this.next = null;
// }

/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */

var addTwoNumbers = function(l1, l2) {
	var result = new ListNode();
	var current = result;
	var	val1, var2, sum,
			carry = 0;
	
	while (l1 || l2) {
		val1 = l1 ? l1.val : 0;
		val2 = l2 ? l2.val : 0;
		sum = val1 + val2 + carry;
		carry = Math.floor(sum / 10);
		
		var newNode = new ListNode(sum % 10);

		if (current.val === undefined) {
			current.val = newNode.val;
		} else {
			current.next = newNode;
			current = newNode;
		}

		if (l1)
			l1 = l1.next;
		if (l2)
			l2 = l2.next;
	}
	if (carry === 1) {
		current.next = new ListNode(1);
	}

	return result;
};
